File: packages/ts/client/src/instructions/withdraw.ts
import { Address, IInstruction, TransactionSigner } from "@solana/kit";
import { ASSOCIATED_TOKEN_PROGRAM_ADDRESS, findAssociatedTokenPda } from "@solana-program/token-2022";
import { TOKEN_PROGRAM_ADDRESS } from "@solana-program/token";
import {
  getCreateAtaInstructions,
  getLendingPositionAddress,
  getLendingVaultAddress,
  getTunaConfigAddress,
  getWithdrawInstruction,
} from "../index.ts";

export async function withdrawInstructions(
  authority: TransactionSigner,
  mint: Address,
  funds: bigint,
  shares: bigint,
): Promise<IInstruction[]> {
  const instructions: IInstruction[] = [];

  // Add create user's token account instruction if needed.
  const createUserAtaInstructions = await getCreateAtaInstructions(
    authority,
    mint,
    authority.address,
    TOKEN_PROGRAM_ADDRESS,
  );
  instructions.push(...createUserAtaInstructions.init);

  // Add withdraw instruction
  const ix = await withdrawInstruction(authority, mint, funds, shares);
  instructions.push(ix);

  // Close WSOL accounts if needed.
  instructions.push(...createUserAtaInstructions.cleanup);

  return instructions;
}

export async function withdrawInstruction(
  authority: TransactionSigner,
  mint: Address,
  funds: bigint,
  shares: bigint,
): Promise<IInstruction> {
  const tunaConfig = (await getTunaConfigAddress())[0];
  const lendingPosition = (await getLendingPositionAddress(authority.address, mint))[0];

  const vault = (await getLendingVaultAddress(mint))[0];
  const vaultAta = (
    await findAssociatedTokenPda({
      owner: vault,
      mint: mint,
      tokenProgram: TOKEN_PROGRAM_ADDRESS,
    })
  )[0];

  const authorityAta = (
    await findAssociatedTokenPda({
      owner: authority.address,
      mint: mint,
      tokenProgram: TOKEN_PROGRAM_ADDRESS,
    })
  )[0];

  return getWithdrawInstruction({
    authority,
    authorityAta,
    lendingPosition,
    mint,
    tunaConfig,
    vault,
    vaultAta,
    associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ADDRESS,
    funds,
    shares,
  });
}

File: rust-sdk/client/src/txbuilder/withdraw.rs
use crate::instructions::{Withdraw, WithdrawInstructionArgs};
use crate::utils::get_create_ata_instructions;
use crate::{get_lending_position_address, get_tuna_config_address, get_vault_address};
use anyhow::Result;
use solana_client::rpc_client::RpcClient;
use solana_instruction::Instruction;
use solana_pubkey::Pubkey;
use spl_associated_token_account::get_associated_token_address_with_program_id;

pub fn withdraw_instructions(rpc: &RpcClient, authority: &Pubkey, mint: &Pubkey, funds: u64, shares: u64) -> Result<Vec<Instruction>> {
    let mint_account = rpc.get_account(mint.into())?;

    let authority_ata_instructions = get_create_ata_instructions(&mint, authority, authority, &mint_account.owner, 0);

    let mut instructions = vec![];
    instructions.extend(authority_ata_instructions.create);
    instructions.push(withdraw_instruction(authority, mint, &mint_account.owner, funds, shares));
    instructions.extend(authority_ata_instructions.cleanup);

    Ok(instructions)
}

pub fn withdraw_instruction(authority: &Pubkey, mint: &Pubkey, token_program: &Pubkey, funds: u64, shares: u64) -> Instruction {
    let tuna_config_address = get_tuna_config_address().0;
    let vault_address = get_vault_address(&mint).0;
    let lending_position_address = get_lending_position_address(&authority, &mint).0;

    let authority_ata = get_associated_token_address_with_program_id(&authority, &mint, token_program);
    let vault_ata = get_associated_token_address_with_program_id(&vault_address, &mint, token_program);

    let ix_builder = Withdraw {
        authority: *authority,
        authority_ata,
        mint: *mint,
        tuna_config: tuna_config_address,
        vault: vault_address,
        vault_ata,
        lending_position: lending_position_address,
        token_program: *token_program,
        memo_program: spl_memo::ID,
    };

    ix_builder.instruction(WithdrawInstructionArgs { funds, shares })
}
